network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py

```diff
@@ -41,6 +62,289 @@ def parse_hardware_block(output_lines, lines):
                 output_lines.append('serial: {}'.format(lines[i].split(' ')[-1]))
 
 
+class CiscoASRXRDeviceInfoPrePostProcessor(PrePostProcessor):
+
+    def parse(self, data):
+        py_logger.info("Parsing output \n{}".format(data))
+        output_lines = []
+        d = dict()
+        lines = data.splitlines()
+        for line in lines:
+            if 'board' in line:
+                d['serial'] = line.split(' ')[-1]
+            if 'uptime' in line:
+                d['name'] = line.split(' ')[0]
+                d['hostname'] = line.split(' ')[0]
+            if 'Software' in line:
+                d['model'] = 'ASR9000'
+        d['os'] = 'IOS XR'
+        d['vendor'] = 'Cisco'
+        d['haState'] = 'ACTIVE'
+        output_lines.append(d)
+        return output_lines
+
+
+class CiscoASRXRInterfacesPrePostProcessor(PrePostProcessor):
+    def parse(self, data):
+
+        name_regex = "(.*) is (administratively )?(up|down), .*"
+        mtu_regex = "MTU (\\d+) bytes.*"
+        ip_regex = "Internet address is (.*)"
+        interface_speed_regex = ".* BW (\\d+) .*"
+        operational_speed_regex = ".*uplex.*, (.*)Mb/s"
+        administrative_status_regex = ".* is (?:administratively )?(up|down), .*"
+        operational_status_regex = ".* line protocol is (up|down)"
+        hardware_address_regex = ".* address is (\\w+\\.\\w+\\.\\w+) .*"
+        duplex_regex = "(.*)-(d|D)uplex.*"
+        vlan_regex = "Encapsulation 802\\.1Q Virtual LAN, Vlan Id\\s+(\\d+).*"
+        active_regex = '(.*).*Active'
+        passive_regex = '(.*).*Passive'
+
+        parser = TextProcessor(SimpleBlockParser())
+        parser.add_rule(Rule(NAME_KEY, name_regex, rule_match_callback))
+        parser.add_rule(Rule(MTU_KEY, mtu_regex, rule_match_callback))
+        parser.add_rule(Rule(IP_KEY, ip_regex, rule_match_callback))
+        parser.add_rule(Rule(IF_SPEED_KEY, interface_speed_regex, rule_match_callback))
+        parser.add_rule(Rule(OP_SPEED_KEY, operational_speed_regex, rule_match_callback))
+        parser.add_rule(Rule(ADMIN_ST_KEY, administrative_status_regex, rule_match_callback))
+        parser.add_rule(Rule(OP_ST_KEY, operational_status_regex, rule_match_callback))
+        parser.add_rule(Rule(HW_KEY, hardware_address_regex, rule_match_callback))
+        parser.add_rule(Rule(DUPLEX_KEY, duplex_regex, rule_match_callback))
+        parser.add_rule(Rule(VLAN_KEY, vlan_regex, rule_match_callback))
+        parser.add_rule(Rule(ACTIVE_PORTS_KEY, active_regex, rule_match_callback))
+        parser.add_rule(Rule(PASSIVE_PORTS_KEY, passive_regex, rule_match_callback))
+        output_lines = parser.process(data)
+        if not bool(output_lines):
+            return []
+
+        for r in output_lines:
+            py_logger.info("Processing row {}".format(r))
+            r.update(duplex=r[DUPLEX_KEY].upper())
+            if r[IF_SPEED_KEY] == '':
+                r[IF_SPEED_KEY] = '0'
```

Comment:
is that required to set 0 ? 
Or can be upload csv with empty values ?

Reply:
It's needed as we are performing mathematical computation over it.